heroku-24/setup.sh

```diff
@@ -0,0 +1,275 @@
+JcHmzisixwS9PMHE+AgPWIGSec/N27a0KmTTvImV6K6nEjXJey0K2+EYJuIBsYUN
+orOGBwDFIhfRk9qGlpgt0KRyguV+AP5qvgry95IrYtrOuE7307SidEbSnvO5ezNe
+mE7gT9Z1tM7IMPfmoKph4BfpNoH7aXiQh1Wo+ChdP92hZUtQrY2Nm13cmkxYjQ4Z
+gMWfYMC+DA/GooSgZM5i6hYqyyfAuUD9kwRN6BqTbuAUAp+hCWYeN4D88sLYpFh3
+paDYNKJ+Gf7Yyi6gThcV956RUFDH3ys5Dk0vDL9NiWwdebWfRFbzoRM3dyGP889a
+OyLzS3mh6nHzZrNGhW73kslSQek8tjKrB+56hXOnb4HaElTZGDvD5wmrrhN94kby
+Gtz3cydIohvNO9d90+29h0eGEDYti7j7maHkBKUAwlcPvMg5m3Y=
+=DA1T
+-----END PGP PUBLIC KEY BLOCK-----
+PGDG_ACCC4CF8
+
+apt-get update
+apt-get upgrade -y --no-install-recommends
+apt-get install -y --no-install-recommends \
+    apt-transport-https \
+    apt-utils \
+    bind9-host \
+    bzip2 \
+    coreutils \
+    curl \
+    dnsutils \
+    ed \
+    file \
+    fontconfig \
+    gcc \
+    geoip-database \
+    gettext-base \
+    gir1.2-harfbuzz-0.0 \
+    git \
+    imagemagick \
+    iproute2 \
+    iputils-tracepath \
+    less \
+    libaom3 \
+    libargon2-1 \
+    libass9 \
+    libc-client2007e \
+    libc6-dev \
+    libcairo2 \
+    libcurl4 \
+    libdatrie1 \
+    libdav1d7 \
+    libev4 \
+    libevent-2.1-7 \
+    libevent-core-2.1-7 \
+    libevent-extra-2.1-7 \
+    libevent-openssl-2.1-7 \
+    libevent-pthreads-2.1-7 \
+    libexif12 \
+    libfreetype6 \
+    libfribidi0 \
+    libgd3 \
+    libgdk-pixbuf2.0-0 \
+    libgdk-pixbuf2.0-common \
+    libgnutls-openssl27 \
+    libgnutls30 \
+    libgraphite2-3 \
+    libgraphite2-3 \
+    libharfbuzz-gobject0 \
+    libharfbuzz-icu0 \
+    libharfbuzz0b \
+    libheif1 \
+    liblzf1 \
+    libmagickcore-6.q16-7-extra \
+    libmcrypt4 \
+    libmemcached11 \
+    libmp3lame0 \
+    libmysqlclient21 \
+    libnuma1 \
+    libogg0 \
+    libonig5 \
+    libopencore-amrnb0 \
+    libopencore-amrwb0 \
+    libopus0 \
+    libpango-1.0-0 \
+    libpangocairo-1.0-0 \
+    libpangoft2-1.0-0 \
+    libpixman-1-0 \
+    librabbitmq4 \
+    librsvg2-2 \
+    librsvg2-common \
+    libsasl2-modules \
+    libseccomp2 \
+    libsodium23 \
+    libspeex1 \
+    libsvtav1enc1d1 \
+    libthai-data \
+    libthai0 \
+    libtheora0 \
+    libunistring5 \
+    libuv1 \
+    libvips42 \
+    libvorbis0a \
+    libvorbisenc2 \
+    libvorbisfile3 \
+    libvpx8 \
+    libwebp7 \
+    libwebpdemux2 \
+    libwebpmux3 \
+    libx264-164 \
+    libx265-199 \
+    libxcb-render0 \
+    libxcb-shm0 \
+    libxrender1 \
+    libxslt1.1 \
+    libyaml-0-2 \
+    libzip4 \
+    libzstd1 \
+    locales \
+    lsb-release \
+    make \
+    netcat-openbsd \
+    openssh-client \
+    openssh-server \
+    patch \
+    poppler-utils \
+    postgresql-client-16 \
+    python-is-python3 \
+    python3 \
+    rename \
+    rsync \
+    shared-mime-info \
+    socat \
+    stunnel \
+    tar \
+    telnet \
+    tzdata \
+    unzip \
+    wget \
+    xz-utils \
+    zip \
+    zlib1g \
+    zstd \
+
+locale-gen en_US.UTF-8
+
+cat > /etc/ImageMagick-6/policy.xml <<'IMAGEMAGICK_POLICY'
```

Comment:
Or alternatively, is it viable for us to drop this custom imagemagick policy entirely? I seem to remember that our custom policy and the default ubuntu imagemagik package's one have mostly converged over the years?

Reply:
I have now in [ad73002](https://github.com/heroku/base-images/pull/245/commits/ad730023b7a94fa017b9d219cf8407481280dda3). Many of the limits were increased. 

The resulting diff after the above changes:

```
--- /etc/ImageMagick-6/policy.xml       2024-02-19 21:38:10.485476018 +0000
+++ /etc/ImageMagick-6/policy.xml.dpkg-new      2023-12-27 10:29:58.000000000 +0000
@@ -1,13 +1,155 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE policymap [
+<!ELEMENT policymap (policy)*>
+<!ATTLIST policymap xmlns CDATA #FIXED "">
+<!ELEMENT policy EMPTY>
+<!ATTLIST policy xmlns CDATA #FIXED "">
+<!ATTLIST policy domain NMTOKEN #REQUIRED>
+<!ATTLIST policy name NMTOKEN #IMPLIED>
+<!ATTLIST policy pattern CDATA #IMPLIED>
+<!ATTLIST policy rights NMTOKEN #IMPLIED>
+<!ATTLIST policy stealth NMTOKEN #IMPLIED>
+<!ATTLIST policy value CDATA #IMPLIED>
+]>
+<!--
+  Creating a security policy that fits your specific local environment
+  before making use of ImageMagick is highly advised. You can find guidance on
+  setting up this policy at https://imagemagick.org/script/security-policy.php,
+  and it's important to verify your policy using the validation tool located
+  at https://imagemagick-secevaluator.doyensec.com/.
+
+
+  Open ImageMagick security policy:
+
+  The default policy for ImageMagick installations is the open security
+  policy. This policy is designed for usage in secure settings like those
+  protected by firewalls or within Docker containers. Within this framework,
+  ImageMagick enjoys broad access to resources and functionalities. This policy
+  provides convenient and adaptable options for image manipulation. However,
+  it's important to note that it might present security vulnerabilities in
+  less regulated conditions. Thus, organizations should thoroughly assess
+  the appropriateness of the open policy according to their particular use
+  case and security prerequisites.
+
+  ImageMagick security policies in a nutshell:
+
+  Domains include system, delegate, coder, filter, module, path, or resource.
+
+  Rights include none, read, write, execute and all.  Use | to combine them,
+  for example: "read | write" to permit read from, or write to, a path.
+
+  Use a glob expression as a pattern.
+
+  Suppose we do not want users to process MPEG video images, use this policy:
+
+    <policy domain="delegate" rights="none" pattern="mpeg:decode" />
+
+  Here we do not want users reading images from HTTP:
+
+    <policy domain="coder" rights="none" pattern="HTTP" />
+
+  The /repository file system is restricted to read only.  We use a glob
+  expression to match all paths that start with /repository:
+
+    <policy domain="path" rights="read" pattern="/repository/*" />
+
+  Prevent users from executing any image filters:
+
+    <policy domain="filter" rights="none" pattern="*" />
+
+  Cache large images to disk rather than memory:
+
+    <policy domain="resource" name="area" value="1GP"/>
+
+  Use the default system font unless overridden by the application:
+
+    <policy domain="system" name="font" value="/usr/share/fonts/favorite.ttf"/>
+
+  Define arguments for the memory, map, area, width, height and disk resources
+  with SI prefixes (.e.g 100MB).  In addition, resource policies are maximums
+  for each instance of ImageMagick (e.g. policy memory limit 1GB, -limit 2GB
+  exceeds policy maximum so memory limit is 1GB).
+
+  Rules are processed in order.  Here we want to restrict ImageMagick to only
+  read or write a small subset of proven web-safe image types:
+
+    <policy domain="delegate" rights="none" pattern="*" />
+    <policy domain="filter" rights="none" pattern="*" />
+    <policy domain="coder" rights="none" pattern="*" />
+    <policy domain="coder" rights="read|write" pattern="{GIF,JPEG,PNG,WEBP}" />
+
+  See https://imagemagick.org/script/security-policy.php for a deeper
+  understanding of ImageMagick security policies.
+-->
 <policymap>
   <policy domain="Undefined" rights="none"/>
+  <!-- Set maximum parallel threads. -->
+  <!-- <policy domain="resource" name="thread" value="2"/> -->
+  <!-- Set maximum time in seconds. When this limit is exceeded, an exception
+       is thrown and processing stops. -->
+  <!-- <policy domain="resource" name="time" value="120"/> -->
+  <!-- Set maximum number of open pixel cache files. When this limit is
+       exceeded, any subsequent pixels cached to disk are closed and reopened
+       on demand. -->
+  <!-- <policy domain="resource" name="file" value="768"/> -->
+  <!-- Set maximum amount of memory in bytes to allocate for the pixel cache
+       from the heap. When this limit is exceeded, the image pixels are cached
+       to memory-mapped disk. -->
   <policy domain="resource" name="memory" value="1024MiB"/>
+  <!-- Set maximum amount of memory map in bytes to allocate for the pixel
+       cache. When this limit is exceeded, the image pixels are cached to
+       disk. -->
   <policy domain="resource" name="map" value="2048MiB"/>
+  <!-- Set the maximum width * height of an image that can reside in the pixel
+       cache memory. Images that exceed the area limit are cached to disk. -->
   <policy domain="resource" name="area" value="256MP"/>
+  <!-- Set maximum amount of disk space in bytes permitted for use by the pixel
+       cache. When this limit is exceeded, the pixel cache is not be created
+       and an exception is thrown. -->
   <policy domain="resource" name="disk" value="2GiB"/>
+  <!-- Set the maximum length of an image sequence.  When this limit is
+       exceeded, an exception is thrown. -->
+  <!-- <policy domain="resource" name="list-length" value="32"/> -->
+  <!-- Set the maximum width of an image.  When this limit is exceeded, an
+       exception is thrown. -->
   <policy domain="resource" name="width" value="32KP"/>
+  <!-- Set the maximum height of an image.  When this limit is exceeded, an
+       exception is thrown. -->
   <policy domain="resource" name="height" value="32KP"/>
-  <policy domain="cache" name="shared-secret" value="passphrase" stealth="true"/>
+  <!-- Periodically yield the CPU for at least the time specified in
+       milliseconds. -->
+  <!-- <policy domain="resource" name="throttle" value="2"/> -->
+  <!-- Do not create temporary files in the default shared directories, instead
+       specify a private area to store only ImageMagick temporary files. -->
+  <!-- <policy domain="resource" name="temporary-path" value="/magick/tmp/"/> -->
+  <!-- Force memory initialization by memory mapping select memory
+       allocations. -->
+  <!-- <policy domain="cache" name="memory-map" value="anonymous"/> -->
+  <!-- Ensure all image data is fully flushed and synchronized to disk. -->
+  <!-- <policy domain="cache" name="synchronize" value="true"/> -->
+  <!-- Replace passphrase for secure distributed processing -->
+  <!-- <policy domain="cache" name="shared-secret" value="secret-passphrase" stealth="true"/> -->
+  <!-- Do not permit any delegates to execute. -->
+  <!-- <policy domain="delegate" rights="none" pattern="*"/> -->
+  <!-- Do not permit any image filters to load. -->
+  <!-- <policy domain="filter" rights="none" pattern="*"/> -->
+  <!-- Don't read/write from/to stdin/stdout. -->
+  <!-- <policy domain="path" rights="none" pattern="-"/> -->
+  <!-- don't read sensitive paths. -->
+  <!-- <policy domain="path" rights="none" pattern="/etc/*"/> -->
+  <!-- Indirect reads are not permitted. -->
   <policy domain="path" rights="none" pattern="@*"/>
+  <!-- These image types are security risks on read, but write is fine -->
+  <!-- <policy domain="module" rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/> -->
+  <!-- This policy sets the number of times to replace content of certain
+       memory buffers and temporary files before they are freed or deleted. -->
+  <!-- <policy domain="system" name="shred" value="1"/> -->
+  <!-- Enable the initialization of buffers with zeros, resulting in a minor
+       performance penalty but with improved security. -->
+  <!-- <policy domain="system" name="memory-map" value="anonymous"/> -->
+  <!-- Set the maximum amount of memory in bytes that are permitted for
+       allocation requests. -->
+  <!-- <policy domain="system" name="max-memory-request" value="256MiB"/> -->
   <policy domain="delegate" rights="none" pattern="URL" />
   <policy domain="delegate" rights="none" pattern="HTTPS" />
   <policy domain="delegate" rights="none" pattern="HTTP" />
```

The main difference between the heroku-24 default and ubuntu:24.04 default now is that heroku-24 has this setting: `<policy domain="cache" name="shared-secret" value="passphrase" stealth="true"/>`, where ubuntu has that commented out. I haven't found much on this. Leaving it felt more secure, and it was fine for heroku-22.

Reply:
> The main difference between the heroku-24 default and ubuntu:24.04 default now is that heroku-24 has this setting: `<policy domain="cache" name="shared-secret" value="passphrase" stealth="true"/>`, where ubuntu has that commented out. I haven't found much on this. Leaving it felt more secure, and it was fine for heroku-22.

Searching the source I found this docs page that for some reason didn't turn up in DDG search:
https://imagemagick.org/script/distribute-pixel-cache.php

And this usage in code:
https://github.com/ImageMagick/ImageMagick6/blob/3da057d31756970cb50b1567922e9caf94d36d84/magick/distribute-cache.c#L242-L263

It seems if anything leaving this option enabled (when it's not enabled in the default config) is less secure, since it's enabling access for distributed processing using a hardcoded password of "passphrase"?!

Reply:
Ok, yeah. I'm not sure where this setting came from, but I'm happy to get rid of it.

Reply:
> Or alternatively, is it viable for us to drop this custom imagemagick policy entirely?

Yeah, this is a better idea. There aren't any differences.

Reply:
I've dropped the imagemagick policy in [2806af8](https://github.com/heroku/base-images/pull/245/commits/2806af80675d3e010e3921456efffcf9d17dc3dc).